src/advanced/time.md

```diff
@@ -0,0 +1,284 @@
+as more flexible.
+
+## Assumptions
+
+The local time on all validator nodes is assumed to be reliable.
+To obtain local reliable time validators can apply external solutions like [tlsdate][tlsdate],
+[roughtime][roughtime], etc.
+
+If the local time on the validator machine is incorrect,
+such node is considered Byzantine. As it is the case with the consensus algorithm,
+the algorithm can tolerate up to 1/3 Byzantine nodes.
+
+## Specification
+
+### Schema
+
+The data schema of the **exonum-time** service consists of two indices:
+
+- **time**  
+  Consolidated time output by the service, which can be used by other business
+  logic on the blockchain.
+- **validators_time**  
+  Merkelized index with the last known local timestamps for all validator nodes.
+  The values in the index are used to update `time` and could be useful
+  for monitoring, diagnostics and the like.
+
+### Transactions
+
+The service implements a single transaction type, allowing a validator to
+output its current time, authenticated by the validator’s digital signature.
+
+The logic of transaction execution is as follows:
+
+1. Check that the transaction signer is one of the validators. If not, quit.
+2. Check that the time specified in the transaction is greater than timestamp
+  of the submitting validator specified in the storage. If not, quit.
+3. Update validator’s time in the `validators_time` index.
+4. If the number of the timestamps in the index belonging to the current validators
+   is at least `2f + 1`, where `f = (n - 1) / 3` is the maximum number
+   of Byzantine validators, then perform the following steps; else quit.
+5. Sort the timestamps by the current validators in the decreasing order
+  (most recent time first).
+7. Take the time with the (1-based) index `f + 1` from the resulting sorted list.
+9. If the taken time `t` is larger than the previous consolidated time,
+  replace the consolidated time with `t`.
+
+Thus, the consolidated time can be updated after each transaction with
+the actual time from any validator node. The procedure takes into account
+possible changes in the validators list, ensures monotony of `time`, and
+is tolerant to the malicious behavior of validator nodes.
+
+## Proof of Correctness
+
+Denote `T` the list of current validators’ timestamps sorted in the decreasing order,
+as specified on step 5 of the algorithm above.
+It is clear that in a system with no more than `f` Byzantine nodes,
+any time from `T` with the (1-based) index in the `[f + 1, 2f + 1]` interval is:
+
+- The time of an honest node, or
+- The time between the timestamps of two honest nodes
+  (and therefore such a time can be considered reliable).
+
+For practical reasons, we always choose the timestamp with index `f + 1`,
+since this value is reliable and at the same time the most recent one.
+
+## Discussion
+
+The validator nodes can potentially generate and send a transaction to update
```

Comment:
"send a transaction" --> "send transactions"
"the nodes send the transaction" --> "the nodes send the transactions"